CMakeLists.txt

```diff
@@ -394,9 +434,11 @@ endif()
 
 if (NOT RWKV_STANDALONE)
     set_property(TARGET ggml PROPERTY GGML_STANDALONE OFF)
-    enable_testing()
-    add_subdirectory(tests)
-    add_subdirectory(extras)
+    if(NOT RWKV_HIPBLAS)
+        enable_testing()
+        add_subdirectory(tests)
+        add_subdirectory(extras)
+    endif()
```

Comment:
Please don't disable tests and extras! Testing is mandatory to verify correctness, especially when adding new backends.

Also, be sure to verify that tests use GPU offloading. See [this line in logit_difference_validator.inc](https://github.com/saharNooby/rwkv.cpp/blob/0df970a6adddd4b938795f92e660766d1e2c1c1f/tests/logit_difference_validator.inc#L32) for reference.

Reply:
Although the dll compiled successfully, the test failed to compile.
The detail is :
```log
FAILED: bin/rwkv_cpu_info.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj extras/CMakeFiles/rwkv_cpu_info.dir/cpu_info.c.obj -o bin\rwkv_cpu_info.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:extras\rwkv_cpu_info.lib -Xlinker /pdb:bin\rwkv_cpu_info.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[11/18] Linking C executable bin\test_context_cloning.exe
FAILED: bin/test_context_cloning.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj tests/CMakeFiles/test_context_cloning.dir/test_context_cloning.c.obj -o bin\test_context_cloning.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:tests\test_context_cloning.lib -Xlinker /pdb:bin\test_context_cloning.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[12/18] Linking C executable bin\test_quantization_format_compatibility.exe
FAILED: bin/test_quantization_format_compatibility.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj tests/CMakeFiles/test_quantization_format_compatibility.dir/test_quantization_format_compatibility.c.obj -o bin\test_quantization_format_compatibility.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:tests\test_quantization_format_compatibility.lib -Xlinker /pdb:bin\test_quantization_format_compatibility.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[13/18] Linking C executable bin\test_quantized_matmul_on_gpu.exe
FAILED: bin/test_quantized_matmul_on_gpu.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj tests/CMakeFiles/test_quantized_matmul_on_gpu.dir/test_quantized_matmul_on_gpu.c.obj -o bin\test_quantized_matmul_on_gpu.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:tests\test_quantized_matmul_on_gpu.lib -Xlinker /pdb:bin\test_quantized_matmul_on_gpu.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)

lld-link: error: undefined symbol: ggml_cuda_transform_tensor
>>> referenced by tests/CMakeFiles/test_quantized_matmul_on_gpu.dir/test_quantized_matmul_on_gpu.c.obj:(main)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[14/18] Linking C executable bin\test_eval_sequence_in_chunks.exe
FAILED: bin/test_eval_sequence_in_chunks.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj tests/CMakeFiles/test_eval_sequence_in_chunks.dir/test_eval_sequence_in_chunks.c.obj -o bin\test_eval_sequence_in_chunks.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:tests\test_eval_sequence_in_chunks.lib -Xlinker /pdb:bin\test_eval_sequence_in_chunks.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[15/18] Linking C executable bin\rwkv_quantize.exe
FAILED: bin/rwkv_quantize.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj extras/CMakeFiles/rwkv_quantize.dir/quantize.c.obj -o bin\rwkv_quantize.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:extras\rwkv_quantize.lib -Xlinker /pdb:bin\rwkv_quantize.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[16/18] Linking C executable bin\test_tiny_rwkv.exe
FAILED: bin/test_tiny_rwkv.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj tests/CMakeFiles/test_tiny_rwkv.dir/test_tiny_rwkv.c.obj -o bin\test_tiny_rwkv.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:tests\test_tiny_rwkv.lib -Xlinker /pdb:bin\test_tiny_rwkv.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[17/18] Linking C executable bin\test_logit_calculation_skipping.exe
FAILED: bin/test_logit_calculation_skipping.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj tests/CMakeFiles/test_logit_calculation_skipping.dir/test_logit_calculation_skipping.c.obj -o bin\test_logit_calculation_skipping.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:tests\test_logit_calculation_skipping.lib -Xlinker /pdb:bin\test_logit_calculation_skipping.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[18/18] Linking C executable bin\test_ggml_basics.exe
FAILED: bin/test_ggml_basics.exe
cmd.exe /C "cd . && C:\PROGRA~1\AMD\ROCm\5.5\bin\clang.exe -fuse-ld=lld-link -nostartfiles -nostdlib -O3 -DNDEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrt -Xlinker /subsystem:console CMakeFiles/ggml.dir/ggml/src/ggml.c.obj CMakeFiles/ggml.dir/ggml/src/ggml-alloc.c.obj tests/CMakeFiles/test_ggml_basics.dir/test_ggml_basics.c.obj -o bin\test_ggml_basics.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:tests\test_ggml_basics.lib -Xlinker /pdb:bin\test_ggml_basics.pdb -Xlinker /version:0.0   rwkv.lib  "C:/Program Files/AMD/ROCm/5.5/lib/hipblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/rocblas.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/clang/17.0.0/lib/windows/clang_rt.builtins-x86_64.lib"  "C:/Program Files/AMD/ROCm/5.5/lib/amdhip64.lib"  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
lld-link: error: undefined symbol: ggml_init_cublas
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_init)

lld-link: error: undefined symbol: ggml_cuda_can_mul_mat
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_graph_plan)

lld-link: error: undefined symbol: ggml_cuda_compute_forward
>>> referenced by CMakeFiles/ggml.dir/ggml/src/ggml.c.obj:(ggml_compute_forward)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Reply:
When I looked at tests/CMakeLists.txt I found
There is this setting:
https://github.com/saharNooby/rwkv.cpp/blob/master/tests/CMakeLists.txt#L4
```
if (GGML_CUDA_SOURCES)
        set_property(TARGET ${TEST_TARGET} PROPERTY CUDA_ARCHITECTURES OFF)
    endif()
```
Does this setting mean that CUDA is ignored?

Reply:
https://cmake.org/cmake/help/latest/prop_tgt/CUDA_ARCHITECTURES.html 
I found the documentation, which is really a bit of a pain as a non-native English speaker